src/trace/statsd.zig

```diff
@@ -0,0 +1,463 @@
+
+        return .{
+            .cluster = cluster,
+            .replica = replica,
+            .implementation = .{
+                .udp = .{
+                    .socket = socket,
+                    .io = io,
+                },
+            },
+            .buffer_completions = buffer_completions,
+            .events_metric_emitted = events_metric_emitted,
+        };
+    }
+
+    // Creates a statsd instance, which will log out the packets that would have been sent. Useful
+    // so that all of the other code can run and be tested in the simulator.
+    pub fn init_log(
+        allocator: std.mem.Allocator,
+        cluster: u128,
+        replica: u8,
+    ) !StatsD {
+        const buffer_completions =
+            try allocator.create(IOPSType(BufferCompletion, packet_count_max));
+        errdefer allocator.destroy(buffer_completions);
+        buffer_completions.* = .{};
+
+        const events_metric_emitted = try allocator.alloc(
+            ?EventMetricAggregate,
+            EventMetric.slot_count,
+        );
+        errdefer allocator.free(events_metric_emitted);
+
+        @memset(events_metric_emitted, null);
+
+        return .{
+            .cluster = cluster,
+            .replica = replica,
+            .implementation = .log,
+            .buffer_completions = buffer_completions,
+            .events_metric_emitted = events_metric_emitted,
+        };
+    }
+
+    pub fn deinit(self: *StatsD, allocator: std.mem.Allocator) void {
+        if (self.implementation == .udp) {
+            self.implementation.udp.io.close_socket(self.implementation.udp.socket);
+        }
+        allocator.free(self.events_metric_emitted);
+        allocator.destroy(self.buffer_completions);
+
+        self.* = undefined;
+    }
+
+    pub fn emit(
+        self: *StatsD,
+        events_metric: []const ?EventMetricAggregate,
+        events_timing: []const ?EventTimingAggregate,
+    ) void {
+        assert(events_metric.len == self.events_metric_emitted.len);
+
+        // This really should not happen; it means we're emitting so many packets, on a short
+        // enough emit timeout, that the kernel hasn't been able to process them all (UDP doesn't
+        // block or provide back-pressure like a TCP socket).
+        //
+        // Keep it as a log, rather than assert, to avoid the common pitfall of metrics killing
+        // the whole system.
+        if (self.buffer_completions.executing() != 0) {
+            log.err("{} / {} packets still in flight; trying to continue", .{
+                self.buffer_completions.executing(),
+                packet_count_max,
+            });
+        }
+
+        if (self.implementation == .udp and self.implementation.udp.send_callback_error_count > 0) {
+            log.warn(
+                "failed to send {} packets",
+                .{self.implementation.udp.send_callback_error_count},
+            );
+            self.implementation.udp.send_callback_error_count = 0;
```

Comment:
Do we need to maintain the error count as a state? Why don't we immediately log error once we receive it? It shouldn't be _too_ many errors I think?

Reply:
I think the idea is to debounce it, so that you don't end up with (in the worst case) 122 logs/second if there is something messed up.

Reply:
(e.g. if you give TB the wrong port, or the stats server is down, you wouldn't want to spam the logs.)

Reply:
Interestingly, if you run the benchmark with `--statsd=...` but don't start any server on that UDP port, despite sending 75-80 packets/s, only 30-40 of those will report errors (`error.ConnectionRefused`)! I guess the others get dropped?